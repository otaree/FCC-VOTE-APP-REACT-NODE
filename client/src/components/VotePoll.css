.VotePoll {
    border: 1px solid rgba(74, 132, 185, 0.733);
    border-radius: 3px;
    width: 60%;
    margin: 2% auto;
    padding: 1% 5%;
    box-sizing: border-box;
}

.VotePoll input[type="radio"] {
    margin: 2%;
}

.VotePoll input[type="radio"]:hover {
    cursor: pointer;
}

.vote_Button {
    width: 10%;
    margin: 1% auto;
}

.vote__btn {
    outline: none;
    border: thin solid #eee;
    border-radius: 3px;
    padding: 10px 15px;
    font-weight: bold;
    font-size: 16px;
    background-color: rgb(71, 148, 88);
}