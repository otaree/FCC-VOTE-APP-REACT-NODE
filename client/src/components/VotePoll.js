import React from 'react';

export default class VotePoll extends React.Component {
    state = {
        selected: null
    };
    handleChange = e => {
        this.setState({ selected: e.target.value });
    };

    handleSubmit = e => {
        e.preventDefault();
        if (!this.state.selected) return;

        this.props.onVote(this.state.selected);
    };
    render() {
        return (
            <div className="VotePoll">
                <form onSubmit={this.handleSubmit}>
                    <div className="level"> 
                        <div className="control level-item">
                        {
                            this.props.poll.options.map(option => {
                                return (
                                        <label key={option._id} className="radio">
                                            <input 
                                                type="radio"
                                                name="vote"
                                                value={option._id}
                                                onChange={this.handleChange}
                                            />
                                             {option.value}
                                        </label>
                                );
                            })
                        }
                        </div>
                    </div>
                    <div className="level">
                        <div className="level-item">
                            <input 
                                className="button is-success"
                                type="submit"
                                value="Vote"
                            />
                        </div>
                    </div>
                </form>
            </div>
        );
    }
}