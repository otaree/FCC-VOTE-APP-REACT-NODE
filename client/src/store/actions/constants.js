// auth
export const AUTH_LOADING = "AUTH_LOADING";
export const AUTH_GET_UID = "AUTH_GET_UID";
export const AUTH_LOGIN = "AUTH_LOGIN";
export const AUTH_LOGOUT = "AUTH_LOGOUT";
export const AUTH_FAIL = "AUTH_FAIL";

// poll
export const POLL_LOADING = "POLL_LOADING";
export const POLL_SUCCESS = "POLL_SUCCESS";
export const POLL_FAIL = "POLL_FAIL";
export const POLL_SET_POLLS = "POLL_SET_POLLS";
export const POLL_SET = "POLL_SET";
export const POLL_UNSET = "POLL_UNSET";