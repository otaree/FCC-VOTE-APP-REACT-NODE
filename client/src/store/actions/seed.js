export const polls = [
    {
        _id: "1",
        question: "Yes or No?",
        createdAt: 1000,
        author: '123abc',
        voters: ["123abc"],
        options: [{
            votes: 1,
            value: 'Yes',
            _id: "101"
        }, {
            votes: 0,
            value: 'No',
            _id: "102"
        }]
    },
    {
        _id: "2",
        question: "2017 or 2018?",
        createdAt: 2000,
        author: '124abc',
        voters: ["124abc"],
        options: [{
            votes: 1,
            value: '2018',
            _id: "201"
        }, {
            votes: 0,
            value: '2017',
            _id: "202"
        }]
    }
];