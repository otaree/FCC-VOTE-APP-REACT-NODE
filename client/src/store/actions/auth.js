import axios from 'axios';
import uniqid from 'uniqid';

import * as constants from './constants';

export const loading = () => {
    return {
        type: constants.AUTH_LOADING,
        loading: true
    };
};

export const login = (user) => {
    return {
        ...user,
        loading: false,
        error: {
            isError: false,
            value: ''
        },
        type: constants.AUTH_LOGIN
    }
};

export const logout = () => {
    return {
        type: constants.AUTH_LOGOUT,
        loading: false,
        error: {
            isError: false,
            value: ''
        }
    };
};

export const fail = (value) => {
    return {
        type: constants.AUTH_FAIL,
        loading: false,
        error: {
            isError: true,
            value
        }
    };
};

export const initUser = (token) => {
    return async dispatch => {
        dispatch(loading());
        try {
            const response = await axios({ url: 'http://localhost:5000/user/token', method: 'get', headers: { 'x-auth': token }});
            dispatch(login(response.data));
        } catch (e) {
            dispatch(fail("Authentication fail"));
        }
    };
};

export const getUid = () => {
    let uid = localStorage.getItem("uid");
    if (!uid) {
        uid = uniqid();
        localStorage.setItem("uid", uid)
    }
    return {
        type: constants.AUTH_GET_UID,
        uid
    }
};