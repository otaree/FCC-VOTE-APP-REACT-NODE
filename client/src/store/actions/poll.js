import axios from 'axios';

import * as constants from './constants';

export const pollLoading = () => ({
    type: constants.POLL_LOADING,
    loading: true
});

export const pollSuccess = () => ({
    type: constants.POLL_SUCCESS,
    loading: false,
    error: {
        isError: false,
        value: ''
    }
});

export const pollFail = value => ({
    type: constants.POLL_FAIL,
    loading: false,
    error: {
        isError: true,
        value
    }
});

export const setPolls = polls => ({
    type: constants.POLL_SET_POLLS,
    loading: false,
    error: {
        isError: false,
        value: ''
    },
    polls
});

export const setPoll = poll => ({
    type: constants.POLL_SET,
    loading: false,
    error: {
        isError: false,
        value: ''
    },
    poll
});

export const fetchPolls = () => {
    return async dispatch => {
        dispatch(pollLoading());

        try {
            const response = await axios.get('http://localhost:5000/polls');
            dispatch(setPolls(response.data));
            return Promise.resolve("PASS");
        } catch (e) {
            dispatch(pollFail("Network Problem"));
            return Promise.reject("FAIL");
        }
    }
};

export const fetchPoll = (id) => {
    return async dispatch => {
        dispatch(pollLoading());

        try {
            const response = await axios.get(`http://localhost:5000/poll/${id}`);
            dispatch(setPoll(response.data));
            return Promise.resolve();
        } catch (e) {
            dispatch(pollFail("Network Problem"));
            return Promise.reject();
        }
    };
};

export const createPoll = (poll, token) => {
    return async dispatch => {
        dispatch(pollLoading());

        try {
            const response = await axios({ url: "http://localhost:5000/poll", method: "post", headers: { "x-auth": token }, data: poll });
            dispatch(pollSuccess());
            return response.data;
        } catch (e) {
            dispatch(pollFail("Network Problem"));
            return Promise.reject();
        }
    }
};

export const updatePoll = (poll, token, id) => {
    return async dispatch => {
        dispatch(pollLoading());

        try {
            const response = await axios({ url: `http://localhost:5000/poll/${id}`, method: "patch", headers: { "x-auth": token }, data: poll });
            dispatch(pollSuccess());
            return response.data;
        } catch (e) {
            dispatch(pollFail("Network Problem"));
            return Promise.reject();
        }
    }
};

export const deletePoll = (token, id) => {
    return async dispatch => {
        dispatch(pollLoading());

        try {
            const response = await axios({ url: `http://localhost:5000/poll/${id}`, method: "delete", headers: { "x-auth": token } });
            dispatch(pollSuccess());
            return response.data;
        } catch (e) {
            dispatch(pollFail("Network Problem"));
            return Promise.reject();
        }
    }
};

export const votePoll = (id, optionId, userId) => {
    return async dispatch => {
        dispatch(pollLoading());

        try {
            const response = await axios({ url: `http://localhost:5000/poll/${id}/vote`, method: "patch", data: { option: optionId, userId } });
            dispatch(pollSuccess());
            return response.data;
        } catch (e) {
            dispatch(pollFail("Network Problem"));
            return Promise.reject();
        }
    }
};