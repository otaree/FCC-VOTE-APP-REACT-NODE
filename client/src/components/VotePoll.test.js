import React from 'react';
import { shallow } from 'enzyme';

import VotePoll from './VotePoll';
import { polls } from '../store/actions/seed';

describe("VotePoll", () => {
    const mockOnVote = jest.fn();
    const props = {
        poll: polls[0],
        onVote: mockOnVote
    };
    let votePoll = shallow(<VotePoll {...props} />);

    it("renders correctly", () => {
        expect(votePoll).toMatchSnapshot();
    });

    describe("when field value changes", () => {
        beforeEach(() => {
            votePoll = shallow(<VotePoll {...props} />);
        });

        it("should set selected value", () => {
            const value = "123abc!";
            votePoll.find('input').at(0).simulate('change', {
                target: { value }
            });

            expect(votePoll.state("selected")).toEqual(value);
        });
    });

    describe("when submitting the form", () => {
        beforeEach(() => {
            votePoll = shallow(<VotePoll {...props}/>);
        });

        it("should correctly submit", () => {
            const value = "123abc!";
            votePoll.setState({ selected: value });
            votePoll.find("form").simulate("submit", {
                preventDefault: () => {}
            });
            expect(mockOnVote).toHaveBeenCalledWith(value);
        });
    });
});