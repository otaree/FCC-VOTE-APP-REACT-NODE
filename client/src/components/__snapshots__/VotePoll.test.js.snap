// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VotePoll renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <VotePoll
    onVote={[Function]}
    poll={
        Object {
            "_id": "1",
            "author": "123abc",
            "createdAt": 1000,
            "options": Array [
              Object {
                "_id": "101",
                "value": "Yes",
                "votes": 1,
              },
              Object {
                "_id": "102",
                "value": "No",
                "votes": 0,
              },
            ],
            "question": "Yes or No?",
            "voters": Array [
              "123abc",
            ],
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <form
        onSubmit={[Function]}
>
        <div
                className="level"
        >
                <div
                        className="control level-item"
                >
                        <label
                                className="radio"
                        >
                                <input
                                        name="vote"
                                        onChange={[Function]}
                                        type="radio"
                                        value="101"
                                />
                                Yes
                        </label>
                        <label
                                className="radio"
                        >
                                <input
                                        name="vote"
                                        onChange={[Function]}
                                        type="radio"
                                        value="102"
                                />
                                No
                        </label>
                </div>
        </div>
        <div
                className="level"
        >
                <div
                        className="level-item"
                >
                        <input
                                className="button is-success"
                                type="submit"
                                value="Vote"
                        />
                </div>
        </div>
</form>,
      "className": "VotePoll",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="level"
>
            <div
                        className="control level-item"
            >
                        <label
                                    className="radio"
                        >
                                    <input
                                                name="vote"
                                                onChange={[Function]}
                                                type="radio"
                                                value="101"
                                    />
                                    Yes
                        </label>
                        <label
                                    className="radio"
                        >
                                    <input
                                                name="vote"
                                                onChange={[Function]}
                                                type="radio"
                                                value="102"
                                    />
                                    No
                        </label>
            </div>
</div>,
          <div
            className="level"
>
            <div
                        className="level-item"
            >
                        <input
                                    className="button is-success"
                                    type="submit"
                                    value="Vote"
                        />
            </div>
</div>,
        ],
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="control level-item"
>
              <label
                            className="radio"
              >
                            <input
                                          name="vote"
                                          onChange={[Function]}
                                          type="radio"
                                          value="101"
                            />
                            Yes
              </label>
              <label
                            className="radio"
              >
                            <input
                                          name="vote"
                                          onChange={[Function]}
                                          type="radio"
                                          value="102"
                            />
                            No
              </label>
</div>,
            "className": "level",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <label
                  className="radio"
>
                  <input
                                    name="vote"
                                    onChange={[Function]}
                                    type="radio"
                                    value="101"
                  />
                  Yes
</label>,
                <label
                  className="radio"
>
                  <input
                                    name="vote"
                                    onChange={[Function]}
                                    type="radio"
                                    value="102"
                  />
                  No
</label>,
              ],
              "className": "control level-item",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": "101",
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <input
                      name="vote"
                      onChange={[Function]}
                      type="radio"
                      value="101"
/>,
                    "Yes",
                  ],
                  "className": "radio",
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "name": "vote",
                      "onChange": [Function],
                      "type": "radio",
                      "value": "101",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": "input",
                  },
                  "Yes",
                ],
                "type": "label",
              },
              Object {
                "instance": null,
                "key": "102",
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <input
                      name="vote"
                      onChange={[Function]}
                      type="radio"
                      value="102"
/>,
                    "No",
                  ],
                  "className": "radio",
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "name": "vote",
                      "onChange": [Function],
                      "type": "radio",
                      "value": "102",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": "input",
                  },
                  "No",
                ],
                "type": "label",
              },
            ],
            "type": "div",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="level-item"
>
              <input
                            className="button is-success"
                            type="submit"
                            value="Vote"
              />
</div>,
            "className": "level",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <input
                className="button is-success"
                type="submit"
                value="Vote"
/>,
              "className": "level-item",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "button is-success",
                "type": "submit",
                "value": "Vote",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            "type": "div",
          },
          "type": "div",
        },
      ],
      "type": "form",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <form
          onSubmit={[Function]}
>
          <div
                    className="level"
          >
                    <div
                              className="control level-item"
                    >
                              <label
                                        className="radio"
                              >
                                        <input
                                                  name="vote"
                                                  onChange={[Function]}
                                                  type="radio"
                                                  value="101"
                                        />
                                        Yes
                              </label>
                              <label
                                        className="radio"
                              >
                                        <input
                                                  name="vote"
                                                  onChange={[Function]}
                                                  type="radio"
                                                  value="102"
                                        />
                                        No
                              </label>
                    </div>
          </div>
          <div
                    className="level"
          >
                    <div
                              className="level-item"
                    >
                              <input
                                        className="button is-success"
                                        type="submit"
                                        value="Vote"
                              />
                    </div>
          </div>
</form>,
        "className": "VotePoll",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="level"
>
              <div
                            className="control level-item"
              >
                            <label
                                          className="radio"
                            >
                                          <input
                                                        name="vote"
                                                        onChange={[Function]}
                                                        type="radio"
                                                        value="101"
                                          />
                                          Yes
                            </label>
                            <label
                                          className="radio"
                            >
                                          <input
                                                        name="vote"
                                                        onChange={[Function]}
                                                        type="radio"
                                                        value="102"
                                          />
                                          No
                            </label>
              </div>
</div>,
            <div
              className="level"
>
              <div
                            className="level-item"
              >
                            <input
                                          className="button is-success"
                                          type="submit"
                                          value="Vote"
                            />
              </div>
</div>,
          ],
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <div
                className="control level-item"
>
                <label
                                className="radio"
                >
                                <input
                                                name="vote"
                                                onChange={[Function]}
                                                type="radio"
                                                value="101"
                                />
                                Yes
                </label>
                <label
                                className="radio"
                >
                                <input
                                                name="vote"
                                                onChange={[Function]}
                                                type="radio"
                                                value="102"
                                />
                                No
                </label>
</div>,
              "className": "level",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <label
                    className="radio"
>
                    <input
                                        name="vote"
                                        onChange={[Function]}
                                        type="radio"
                                        value="101"
                    />
                    Yes
</label>,
                  <label
                    className="radio"
>
                    <input
                                        name="vote"
                                        onChange={[Function]}
                                        type="radio"
                                        value="102"
                    />
                    No
</label>,
                ],
                "className": "control level-item",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": "101",
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      <input
                        name="vote"
                        onChange={[Function]}
                        type="radio"
                        value="101"
/>,
                      "Yes",
                    ],
                    "className": "radio",
                  },
                  "ref": null,
                  "rendered": Array [
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "name": "vote",
                        "onChange": [Function],
                        "type": "radio",
                        "value": "101",
                      },
                      "ref": null,
                      "rendered": null,
                      "type": "input",
                    },
                    "Yes",
                  ],
                  "type": "label",
                },
                Object {
                  "instance": null,
                  "key": "102",
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      <input
                        name="vote"
                        onChange={[Function]}
                        type="radio"
                        value="102"
/>,
                      "No",
                    ],
                    "className": "radio",
                  },
                  "ref": null,
                  "rendered": Array [
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "name": "vote",
                        "onChange": [Function],
                        "type": "radio",
                        "value": "102",
                      },
                      "ref": null,
                      "rendered": null,
                      "type": "input",
                    },
                    "No",
                  ],
                  "type": "label",
                },
              ],
              "type": "div",
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <div
                className="level-item"
>
                <input
                                className="button is-success"
                                type="submit"
                                value="Vote"
                />
</div>,
              "className": "level",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <input
                  className="button is-success"
                  type="submit"
                  value="Vote"
/>,
                "className": "level-item",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "button is-success",
                  "type": "submit",
                  "value": "Vote",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              "type": "div",
            },
            "type": "div",
          },
        ],
        "type": "form",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
